#00.PyTorch Fundamentals Exercises
##1. Documentation reading
A big part of deep learning (and learning to code in general) is getting familiar with the documentation of a certain framework you're using.  We'll be using PyTorch documentation a lot throughout the rest of this course. So I'd recommend spedning 10-minutes reading the following (it's okay if you don't get some things for now, the focus is not yet full understanding, it's awareness):

- The documentation on `torch.Tensor': https://pytorch.org/docs/stable/tensors.html#torch-tensor

- The documentation on 'torch.cuda': https://pytorch.org/docs/master/notes/cuda.html#cuda-semantics

In [ ]:
# No code as the assignment is to read the documentation in the links

#2. Create a random tensor with shape (7,7)

In [5]:
# Import torch
import torch

# Create random tensor
T1_rand = torch.rand(7,7)
print(T1_rand)

tensor([[0.6645, 0.7183, 0.8554, 0.7040, 0.5511, 0.4463, 0.0911],
        [0.2174, 0.3646, 0.0917, 0.8927, 0.5832, 0.4595, 0.4348],
        [0.3601, 0.0011, 0.4529, 0.6208, 0.1518, 0.9934, 0.5268],
        [0.3578, 0.2817, 0.8870, 0.6805, 0.8788, 0.6338, 0.4643],
        [0.3128, 0.9499, 0.3349, 0.6560, 0.5388, 0.4328, 0.9804],
        [0.8186, 0.0669, 0.8932, 0.6890, 0.1348, 0.3924, 0.5374],
        [0.0077, 0.5913, 0.1057, 0.0561, 0.0583, 0.0074, 0.0239]])


#3. Perform a matrix multiplication on the tensor from 2 with another random tensor with shape (1,7) (hint: you may need to transpose the second tensor).

In [10]:
# Create another random tensor
T2_rand = torch.rand(1,7)
T2_rand

# Perform matrix multiplication
T3_matmult = torch.matmul(T1_rand, T2_rand.T)
T3_matmult

tensor([[1.9794],
        [1.3091],
        [1.1908],
        [1.6349],
        [2.2597],
        [1.6014],
        [0.6414]])

#4. Set the random seed to `0` and do 2 & 3 over again

The output should be:

(tensor([[1.8542],
         [1.9611],
         [2.2884],
         [3.0481],
         [1.7067],
         [2.5290],
         [1.7989]]), torch.Size([7, 1]))

In [13]:
# Set the manual seed
RANDOM_SEED = 0
torch.manual_seed(RANDOM_SEED)

T1_randS0 = torch.rand(7,7)
print(f"T1_randS0 = {T1_randS0}")

T2_randS0 = torch.rand(1,7)
print(f"T2_randS0 = {T2_randS0}")

T3_matmultS0 = torch.matmul(T1_randS0, T2_randS0.T)
print(f"T3_matmultS0 = {T3_matmultS0}")

T1_randS0 = tensor([[0.4963, 0.7682, 0.0885, 0.1320, 0.3074, 0.6341, 0.4901],
        [0.8964, 0.4556, 0.6323, 0.3489, 0.4017, 0.0223, 0.1689],
        [0.2939, 0.5185, 0.6977, 0.8000, 0.1610, 0.2823, 0.6816],
        [0.9152, 0.3971, 0.8742, 0.4194, 0.5529, 0.9527, 0.0362],
        [0.1852, 0.3734, 0.3051, 0.9320, 0.1759, 0.2698, 0.1507],
        [0.0317, 0.2081, 0.9298, 0.7231, 0.7423, 0.5263, 0.2437],
        [0.5846, 0.0332, 0.1387, 0.2422, 0.8155, 0.7932, 0.2783]])
T2_randS0 = tensor([[0.4820, 0.8198, 0.9971, 0.6984, 0.5675, 0.8352, 0.2056]])
T3_matmultS0 = tensor([[1.8542],
        [1.9611],
        [2.2884],
        [3.0481],
        [1.7067],
        [2.5290],
        [1.7989]])


#5. Speaking of random seeds, we saw how to set it with `torch.manual_seed()` but is there a GPU equivalent? (hint: you'll need to look into the documentation for `torch.cuda` for this one)

- If there is, set the GPU random seed to `1234`.

In [29]:
# Set random seed on GPU
import torch

RANDOM_SEED_GPU = 1234
RANDOM_SEED_CPU = 42

torch.cuda.is_available()

device = "cuda" if torch.cuda.is_available() else "cpu"

if torch.cuda.is_available():
  torch.manual_seed(RANDOM_SEED_GPU)
else:
  torch.use_deterministic_algorithms(True)
  torch.manual_seed(RANDOM_SEED_CPU)

print(device)
print(torch.rand(2))

cpu
tensor([0.8823, 0.9150])


#6. Create two random tensors of shape `(2, 3)` and send them both to the GPU (you'll need access to a GPU for this).  Set `torch.manual_seed(1234)` when creating the tensors (this doesn't have to be the GPU random seed).  The output should be something like:

Device: cuda
(tensor([[0.0290, 0.4019, 0.2598],
         [0.3666, 0.0583, 0.7006]], device='cuda:0'),
 tensor([[0.0518, 0.4681, 0.6738],
         [0.3315, 0.7837, 0.5631]], device='cuda:0'))

In [56]:
# Set the random seed
import torch

RANDOM_SEED_GPU = 1234
RANDOM_SEED_CPU = 42

# Check for access to GPU
torch.cuda.is_available()

device = "cuda" if torch.cuda.is_available() else "cpu"

if torch.cuda.is_available():
  torch.manual_seed(RANDOM_SEED_GPU)
  print("GPU is available!")
  print("")

  # Create two random tensors on GPU
  T1 = torch.rand(2,3)
  T2 = torch.rand(2,3)
  T1 = T1.to(device)
  T2 = T2.to(device)
else:
  print("GPU not available")
  print("")
  torch.use_deterministic_algorithms(True)
  torch.manual_seed(RANDOM_SEED_CPU)
  T1 = torch.rand(2,3)
  T2 = torch.rand(2,3)

print(f"T1 tensor = {T1} on {device}")
print("")
print(f"T2 tensor = {T2} on {device}")




GPU not available

T1 tensor = tensor([[0.8823, 0.9150, 0.3829],
        [0.9593, 0.3904, 0.6009]]) on cpu

T2 tensor = tensor([[0.2566, 0.7936, 0.9408],
        [0.1332, 0.9346, 0.5936]]) on cpu


#7. Perform a matrix multiplication on the tensors you created in 6 (again, you may have to adjust the shapes of the tensors.

The output should look like:

(tensor([[0.3647, 0.4709],
         [0.5184, 0.5617]], device='cuda:0'), torch.Size([2, 2]))

In [47]:
# Perform matmul on tensors T1 and T2
T3 = torch.matmul(T1, T2.T)
print(f"Matrix Multiply of T1 and T2 = {T3} calculated on {device}, {T3.shape}")

Matrix Multiply of T1 and T2 = tensor([[1.3127, 1.1999],
        [1.1213, 0.8494]]) calculated on cpu, torch.Size([2, 2])


#8. Find the maximum and minimum values of the output of 7.

In [49]:
# Find max
print(f"The maximum value of T3 = {torch.max(T3)}")

# Find min
print(f"The minimum value of T3 = {torch.min(T3)}")

The maximum value of T3 = 1.312738299369812
The minimum value of T3 = 0.8493574261665344


#9. Find the index values for the maximum and minimum values of the output of 7.

In [52]:
# Find the arg max
print(f"The index of T3 for the maximum value = {T3.argmax()}")

# Find the arg min
print(f"The index of T3 for the minimum value = {T3.argmin()}")

The index of T3 for the maximum value = 0
The index of T3 for the minimum value = 3


#10. Make a random tensor with shape `(1, 1, 1, 10)` and then create a new tensor with all the `1` dimensions removed to be left with a tensor of shape `10`.  Set the seed to `7` when you create it an print out the first tensor and it's shape as well as the second tensor and it's shape.

In [55]:
# Set the seed
RANDOM_SEED = 7
torch.manual_seed(RANDOM_SEED)

# Create random tensor
T4 = torch.rand([1,1,1,10])
print(f"Random tensor T4 = {T4}, with shape of {T4.shape}")

# Remove single dimensions
T5 = T4.squeeze()
print("")
print(f"T4 tensor squeezed to make T5 = {T5}, with shape of {T5.shape}")

Random tensor T4 = tensor([[[[0.5349, 0.1988, 0.6592, 0.6569, 0.2328, 0.4251, 0.2071, 0.6297,
           0.3653, 0.8513]]]]), with shape of torch.Size([1, 1, 1, 10])

T4 tensor squeezed to make T5 = tensor([0.5349, 0.1988, 0.6592, 0.6569, 0.2328, 0.4251, 0.2071, 0.6297, 0.3653,
        0.8513]), with shape of torch.Size([10])
